packages/shared/src/store/sync.ts

```diff
@@ -268,6 +269,44 @@ export const syncSystemContacts = async (ctx?: SyncCtx) => {
   }
 };
 
+export const syncContactDiscovery = async (ctx?: SyncCtx) => {
+  const systemContacts = await api.getSystemContacts();
+  const phoneNumbers = systemContacts
+    .map((contact) => contact.phoneNumber)
+    .filter((phoneNumber) => phoneNumber && phoneNumber.length > 0) as string[];
+
+  if (!phoneNumbers.length) {
+    logger.trackEvent(AnalyticsEvent.DebugSystemContacts, {
+      context: 'no phone numbers found, skipping discovery',
+    });
+    // this should also mean we no-op on web since we don't have any
+    // system contacts
+    return;
+  }
+
+  try {
+    const matches = (
+      await syncQueue.add('discoverContacts', ctx, () =>
+        api.discoverContacts(phoneNumbers)
+      )
+    ).filter((match) => match[1] !== api.getCurrentUserId());
+    logger.log('got contact discovery matches', matches);
+
+    await db.linkSystemContacts({ matches });
+    logger.log('inserted contact discovery matches', matches);
+    const contactIds = matches.map((m) => m[1]);
+    await addContacts(contactIds);
```

Comment:
can we add a separate error contexts for api call failure vs add contacts failure? Would be helpful to distinguish between the two if we have to debug

Reply:
we may also want to distinguish `db.linkSystemContacts` failures since that logic is complicated